### PR TITLE
Consolidate common code in cts-exec and cts-fencing

### PR DESCRIPTION
**Pull request by clumens**
Here's the followup to the initial python PR.  This is all private API, condensing cts-exec and cts-fencing into library calls that should be sufficient for writing cts-attrd.  There's likely much more that can be done.  For instance, I've only glanced at cts-regression and cts-scheduler enough to see that they can't use any of these new library functions.  Hopefully this opens it up enough to where more people can contribute.

### CODE REVIEW

File: cts/cts-exec.in [hunk 1]
```diff
@@ -988,28 +988,28 @@ class TestOptions(object):
...
+                "\n\t\tDefaults to 2. The value 0 means no limit.")
+        print("\t [--force-wait | -w]"
+                "\n\t\tEach test case waits the default/specified --timeout for the daemon without tracking the log.")
+        if BuildOptions.REMOTE_ENABLED:
+            print("\t [--pacemaker-remote | -R             Test pacemaker-remoted binary instead of pacemaker-execd")
+        print("\t [--run-only-pattern | -p 'string']   Run only tests containing the string value")
+        print("\n\tExample: Run only the test 'start_stop'")
+        print("\t\t " + sys.argv[0] + " --run-only start_stop")
+        print("\n\tExample: Run only the tests with the string 'systemd' present in them")
+        print("\t\t " + sys.argv[0] + " --run-only-pattern systemd")
```
Comment (kgaillot) [inline cts/cts-exec.in:946]: Not for this PR, but eventually I'd like to move to the argparse module for all Python executables, which will get rid of a lot of custom arg parsing and allow us to share some argument definitions across executables

File: python/pacemaker/_cts/test.py [hunk 2]
```diff
@@ -284,7 +294,10 @@ def run_cmd(self, args):
             ### launched earlier that is listening for the failure may not have
             ### time to connect to pacemaker-execd.
             time.sleep(2)
-            subprocess.Popen(shlex.split(args['kill']))
+            # If the kill process exits with non-zero, check=False will prevent
+            # an exception from being raised.  Basically, we'll just ignore the
+            # failure.
+            subprocess.run(shlex.split(args['kill']), check=False)
```
Comment (kgaillot) [inline python/pacemaker/_cts/test.py:300]: We're almost there but not quite :)  subprocess.run was added in Python 3.5

**Review by kgaillot — COMMENTED** [review]
Beautiful :)

File: python/pacemaker/_cts/test.py [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+        return self._result_exitcode
+
+    @exitcode.setter
+    def exitcode(self, value):
+        self._result_exitcode = value
+
+    @property
+    def logpath(self):
+        """ The path to the log for whatever daemon is being tested.  Note that
+            this requires all subclasses to sef self._daemon_location before
```
[the remark this answers is not in the document]
Comment (clumens) [reply python/pacemaker/_cts/test.py:113]: Fixed.

File: python/pacemaker/_cts/test.py (hunk 2 above)
Comment (clumens) [reply python/pacemaker/_cts/test.py:300]: Weirdly, the pylint version I have installed now does not complain about using Popen here while a previous version did.  That's why I moved to using run.  Fixed.

File: cts/cts-fencing.in [hunk 4]
```diff
@@ -28,7 +28,7 @@ if os.path.exists("@abs_top_builddir@/python") and "@abs_top_builddir@" != "@abs
 from pacemaker.buildoptions import BuildOptions
 from pacemaker.exitstatus import ExitStatus
 from pacemaker._cts.errors import ExitCodeError, OutputFoundError, OutputNotFoundError, XmlValidationError
-from pacemaker._cts.process import killall, exit_if_proc_running
+from pacemaker._cts.process import *
```
[the remark this answers is not in the document]
Comment (clumens) [reply cts/cts-fencing.in:29]: I like the explicit imports too.  I must have been feeling lazy when I wrote this.

File: cts/cts-exec.in [hunk 5]
```diff
@@ -62,47 +63,32 @@ def update_path():
...
-        self.cmd_tool_output = ""
-        self.result_exitcode = ExitStatus.OK
-
-        self.execd_process = None
-        self.stonith_process = None
+            self._daemon_location = "pacemaker-execd"
 
-        self.executed = False
+        self._execd_process = None
+        self._test_tool_location = "cts-exec-helper"
```
[the remark this answers is not in the document]
Comment (clumens) [reply cts/cts-exec.in:78]: stonith_process should be fine - it's set to None in `Test.__init__` and then to something later on in start_environment if self.tls is False.  I think that's how it used to be.  logpath/logdir should likewise be fine - logdir is set in `Test.__init__` and then logpath is a property.

File: cts/cts-exec.in [hunk 6]
```diff
@@ -179,27 +173,28 @@ class ExecTest(Test):
...
 
-        self._new_cmd(cmd, args, ExitStatus.OK, "")
+        self._new_cmd(cmd, args, ExitStatus.OK)
 
     def add_cmd_check_stdout(self, args, match, no_match=""):
         """ Add a command with expected output to be executed as part of this test """
 
-        self._new_cmd(self._test_tool_location, args, ExitStatus.OK, match, False, no_match)
+        self._new_cmd(self._test_tool_location, args, ExitStatus.OK,
+                      stdout_match=match, stdout_megative_match=no_match)
```
[the remark this answers is not in the document]
Comment (clumens) [reply cts/cts-exec.in]: I like megative more, but fixed.

File: cts/cts-exec.in (hunk 5 above)
[the remark this answers is not in the document]
Comment (kgaillot) [reply cts/cts-exec.in:78]: Ah cool, I overlooked those somehow.

**Comment by clumens** [comment]
Hopefully last push of this.  At the least, I don't have anything else to do right now.

**Review by kgaillot — APPROVED** [review]
Awesome :)

**Pull request merged**